legion/engine/llri/detail/command_group.hpp

```diff
@@ -0,0 +1,141 @@
+         *
+         * @param desc The allocation description describes the command list's usage. Each CommandList is allocated with this same structure.
+         * @param count The number of CommandLists to allocate.
+         * @param cmdLists A pointer to the resulting command list vector.
+         *
+         * @return Success upon correct execution of the operation.
+         * @return ErrorInvalidUsage if cmdLists was nullptr.
+         * @return ErrorInvalidUsage if desc.usage was not a valid command_list_usage enum value.
+         * @return ErrorInvalidUsage if count was 0.
+         * @return ErrorExceededLimit if the CommandGroup doesn't have enough CommandLists to allocate count CommandLists.
+         * @return Implementation defined result values: ErrorOutOfHostMemory, ErrorOutOfDeviceMemory.
+        */
+        result allocate(const command_list_alloc_desc& desc, uint8_t count, std::vector<CommandList*>* cmdLists);
+
+        /**
+         * @brief Free the CommandList, which allows the CommandGroup to release all of the device memory allocated for the CommandList.
+         *
+         * @param cmdList The CommandList to be free'd and destroyed.
+         *
+         * @return Success upon correct execution of the operation.
+         * @return ErrorInvalidUsage if cmdList is nullptr.
+         * @return ErrorInvalidUsage if the cmdList didn't belong to the CommandGroup.
+        */
+        result free(CommandList* cmdList);
+
+        /**
+         * @brief Free an array of CommandLists, which allows the CommandGroup to release all of the device memory allocated for the CommandLists.
+         *
+         * @param numCommandLists The number of CommandLists to be free'd, *should** match the size of the cmdLists array.
+         * @param cmdLists An array of CommandGroups, **must** not be nullptr and **must** at least contain numCommandLists amount of CommandLists.
+         *
+         * @return Success upon correct execution of the operation.
+         * @return ErrorInvalidUsage if numCommandLists is 0.
+         * @return ErrorInvalidUsage if cmdLists is nullptr.
+         * @return ErrorInvalidUsage if any of the CommandLists in cmdLists is nullptr.
+         * @return ErrorInvalidUsage if any of the CommandLists in cmdLists doesn't belong to the CommandGroup.
+        */
+        result free(uint8_t numCommandLists, CommandList** cmdLists);
+
+    private:
+        //Force private constructor/deconstructor so that only create/destroy can manage lifetime
+        CommandGroup() = default;
+        ~CommandGroup() = default;
+
+        void* m_ptr = nullptr;
+
+        void* m_deviceHandle = nullptr;
+        void* m_deviceFunctionTable = nullptr;
+
+        validation_callback_desc m_validationCallback;
+        void* m_validationCallbackMessenger = nullptr;
+
+        queue_type m_type;
+        uint8_t m_maxCount;
+        std::list<CommandList*> m_cmdLists;
```

Comment:
Is there a particular reason to use std::list here? since std::unordered_set has an O(1) insertion and deletion, while std::list has O(1) insertion and O(n) deletion.

Reply:
I assumed std::list would be the fastest at adding/removing but if unordered_set is faster I'll switch to that